### PR TITLE
Feature/new types

### PR DESCRIPTION
**Pull request by ZharaOo**
**Examples of test classes with new features**
**Updated**
[ClassesExamples.zip](https://github.com/TouchInstinct/api-generator-templates/files/3191028/ClassesExamples.zip)

### CODE REVIEW

File: Swift/blocks/class/copy-declaration.twig [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+{%- set containsOptionalFields = false -%}
+{%- if allFieldsOrdered is not empty -%}
+    {%- for field in allFieldsOrdered -%}
+        {%- if field.nullable or field.optional -%}
+            {%- set containsOptionalFields = true -%}
+        {%- endif -%}
+    {%- endfor -%}
+{%- endif -%}
+
+{% if containsOptionalFields %}
```
Comment (petropavel13) [inline Swift/blocks/class/copy-declaration.twig:15]: we need to be able to copy with new parameters even if there is an optional fields

File: Swift/blocks/class/copy-declaration.twig [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{%- import '../../macroses/common.utils.twig' as utils -%}
+
+    func copy{%- if hasChilds -%}{{ type.baseTypeName }}{%- endif -%}With({%- include '../class/nullable-parameters-fields.twig' with { fields: allFieldsOrdered } -%}) -> {{ type.baseTypeName }} {
+        return {{ type.baseTypeName }}({% include '../class/fields-optional-initialization.twig' with { fields: allFieldsOrdered } %})
+    }
+
+{%- set containsOptionalFields = false -%}
```
Comment (petropavel13) [inline Swift/blocks/class/copy-declaration.twig]: seems that is redundant

File: Swift/macroses/common.utils.twig [hunk 3]
```diff
@@ -10,19 +10,36 @@
...
 
+{% macro numberFormatterFrom(fields) %}
+    {%- set containsStringDecimal = false -%}
+    {%- for field in fields -%}
+        {%- if field.type.type.baseTypeName == "StringDecimal" -%}
+            {%- set containsStringDecimal = true -%}
+        {%- endif -%}
+    {%- endfor -%}
+    {%- if containsStringDecimal -%}
+        let formatter = NumberFormatter()
```
Comment (petropavel13) [inline Swift/macroses/common.utils.twig]: for performance purpose it should be rendered like ApiDateFormattingService

File: Swift/macroses/complexField.utils.twig [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        {%- set dateTimeInit = 'ApiDateFormattingService.date(from: {{ field.name }}, format: .{{ dateFormatToName(field.type.dateFormat) }}, parsedIn: nil)' -%}
+        {{ self.decodeThrowableField(field, dateTimeInit, 'init?(string:format:fromRegion:) returned nil') -}}
+    {%- elseif field.type.type.baseTypeName == "Color" -%}
+        {{ self.decodeThrowableField(field, 'UIColor(hexString: {{ field.name }})', 'Unable to decode hex string') -}}
+    {%- elseif field.type.type.baseTypeName == "Url" -%}
+        {{ self.decodeThrowableField(field, 'URL(string: {{ field.name }})', 'Unable to decode URL from string') -}}
+    {%- elseif field.type.type.baseTypeName == "StringDecimal" -%}
+        {{- self.decodeThrowableField(field, 'formatter.number(from: {{ field.name }})', 'Unable to decode decimal from string') -}}
+    {%- elseif field.type.type.baseTypeName == "DateTimeTimestamp" -%}
+        self.{{ field.name }} = DateInRegion(seconds: TimeInterval({{ field.name }}), region: .local)
```
Comment (petropavel13) [inline Swift/macroses/complexField.utils.twig]: don't pass region. timestamp, usually, uses UTC

File: Swift/macroses/common.utils.twig [hunk 5]
```diff
@@ -120,56 +142,12 @@
...
-    {%- endif -%}
-{% endmacro %}
-
 {% macro enumValueName(value) %}
     {%- import _self as self -%}
 
     {{- self.decapitalize(value.name) -}}
 {%- endmacro %}
 
-{% macro defaultValueForField(field) %}
```
Comment (petropavel13) [inline Swift/macroses/common.utils.twig]: revert please, we need default values for non-optional types

**Review by petropavel13 — CHANGES_REQUESTED** [review]

File: Swift/blocks/class/copy-declaration.twig (hunk 1 above)
Comment (ZharaOo) [reply Swift/blocks/class/copy-declaration.twig:15]: To copy object with new parameters we can use copyWith method declared above in this file

Comment (petropavel13) [reply Swift/blocks/class/copy-declaration.twig:15]: maybe you mean `if !containsOptionalFields` ?

File: Swift/ApiNumberFormattingService.swift.twig [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+enum ApiNumberFormat: NumberFormat {
+    
+    case decimal
+
+    var numberFormatter: NumberFormatter {
+        let numberFormatter = NumberFormatter()
+
+        switch self {
+        case .decimal:
```
Comment (petropavel13) [inline Swift/ApiNumberFormattingService.swift.twig]: switch is useless for now

File: Swift/macroses/common.utils.twig [hunk 7]
```diff
@@ -34,21 +38,38 @@
...
+    {%- elseif field.type.type.baseTypeName == "Url" -%}
+        {{ self.formatStrongOrNullableOptional(field, strong) -}}.absoluteString
+    {%- elseif field.type.type.baseTypeName == "StringDecimal" -%}
+        {{- self.formatEncodingStringDecimal(field) -}}
     {%- else -%}
         {{ field.name }}
     {%- endif -%}
 {% endmacro %}
 
+{% macro formatStrongOrNullableOptional(field, strong) %}
```
Comment (petropavel13) [inline Swift/macroses/common.utils.twig]: I don't understand what does mean `strong` here. Can you rename it to something more clear?

File: Swift/macroses/common.utils.twig [hunk 8]
```diff
@@ -34,21 +38,38 @@
...
+        {{ self.formatNullableOrOptional(field.name, field.nullable, field.optional) -}}
+    {%- endif -%}
+{% endmacro %}
+
 {%- macro formatEncodingDate(field) -%}
     ApiDateFormattingService.string(from: {{ field.name -}}, format: .{{- dateFormatToName(field.type.dateFormat) -}})
 {%- endmacro -%}
 
+{%- macro formatEncodingStringDecimal(field) -%}
+    ApiNumberFormattingService.shared.string(from: {{ field.name -}}, format: .decimal)
```
Comment (petropavel13) [inline Swift/macroses/common.utils.twig]: shared is required?

File: Swift/macroses/complexField.utils.twig [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        {%- set dateTimeInit = "ApiDateFormattingService.date(from: %s, format: .%s, parsedIn: nil)"|format(field.name, field.type.dateFormat) -%}
+        {{ self.decodeThrowableField(field, dateTimeInit, 'init?(string:format:fromRegion:) returned nil') -}}
+    {%- elseif field.type.type.baseTypeName == "Color" -%}
+        {%- set colorInit = "UIColor(hexString: %s)"|format(field.name) -%}
+        {{ self.decodeThrowableField(field, colorInit, 'Unable to decode hex string') -}}
+    {%- elseif field.type.type.baseTypeName == "Url" -%}
+        {%- set urlInit = "URL(string: %s)"|format(field.name) -%}
+        {{ self.decodeThrowableField(field, urlInit, 'Unable to decode URL from string') -}}
+    {%- elseif field.type.type.baseTypeName == "StringDecimal" -%}
+        {%- set stringDecimalInit = "ApiNumberFormattingService.shared.decimalNumber(from: %s, format: .decimal)"|format(field.name) -%}
```
Comment (petropavel13) [inline Swift/macroses/complexField.utils.twig]: shared is required?

**Review by petropavel13 — CHANGES_REQUESTED** [review]

File: Swift/blocks/class/copy-declaration.twig (hunk 1 above)
Comment (ZharaOo) [reply Swift/blocks/class/copy-declaration.twig:15]: That is a declaration of 'copyWithout' method. We decided that we can copy a class only without optional fields. So if class doesn't contains optional fields we don't need to declare this method.

**Review by petropavel13 — APPROVED** [review]

**Pull request merged**